### PR TITLE
Fix outputf by making it a macro

### PR DESCRIPTION
**Pull request by johnboiles**

### CODE REVIEW

File: blockware/lib/Text/Text.h [hunk 1]
```diff
@@ -1,6 +1,10 @@
 #include <Colors.h>
+#include <Adafruit_SSD1351.h>
+
+extern Adafruit_SSD1351 tft;
+
+#define outputf(color, size, msg, ...) do {tft.setTextSize(size); tft.setTextColor(color); tft.printf(msg, ##__VA_ARGS__);} while (0)
```
Comment (magus) [inline blockware/lib/Text/Text.h]: i like this a lot, we have to wrap it in a do-while because it needs to be block scoped?

**Comment by magus** [comment]
Build error looks weird, not sure what is wrong but looks like it's in the library code not ours

File: blockware/lib/Text/Text.h (hunk 1 above)
Comment (johnboiles) [reply blockware/lib/Text/Text.h]: I forget why this is good practice but I seem to remember it is. I think there's some weird edge case with syntax if you don't do this.

**Comment by johnboiles** [comment]
yeah it's weird too that the branch build passed but not the PR build

**Comment by johnboiles** [comment]
Maybe it's something about the order the libraries are built?

```
make -C blockware/homeassistant-display test
make[1]: Entering directory '/home/travis/build/johnboiles/blocks-with-screens/blockware/homeassistant-display'
cp src/secrets.h.example src/secrets.h
pio run
Processing esp12e (platform: espressif8266; board: esp12e; framework: arduino)
--------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/espressif8266/esp12e.html
PLATFORM: Espressif 8266 2.2.3 > Espressif ESP8266 ESP-12E
HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
PACKAGES: toolchain-xtensa 2.40802.190218 (4.8.2), tool-esptool 1.413.0 (4.13), tool-esptoolpy 1.20600.0 (2.6.0), framework-arduinoespressif8266 2.20502.0 (2.5.2)
LDF: Library Dependency Finder -> http://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Looking for Adafruit GFX Library library in registry
Found: https://platformio.org/lib/show/13/Adafruit GFX Library
LibraryManager: Installing id=13
Using cache: /home/travis/.platformio/.cache/f3/65928c0469cb491eff3b3e8548cba2f3
Unpacking...
Adafruit GFX Library @ 1.5.7 has been successfully installed!
Looking for Adafruit SSD1351 library library in registry
Found: https://platformio.org/lib/show/266/Adafruit SSD1351 library
LibraryManager: Installing id=266
Using cache: /home/travis/.platformio/.cache/a1/7334df3a4b91c6ca4d2729054ea9d0a1
Unpacking...
Adafruit SSD1351 library @ 1.2.2 has been successfully installed!
Looking for HARestAPI library in registry
Found: https://platformio.org/lib/show/5828/HARestAPI
LibraryManager: Installing id=5828
Downloading...
Unpacking...
HARestAPI @ 0.2 has been successfully installed!
Looking for ArduinoJson library in registry
Found: https://platformio.org/lib/show/64/ArduinoJson
LibraryManager: Installing id=64 @ 6.12.0
Downloading...
Unpacking...
ArduinoJson @ 6.12.0 has been successfully installed!
Looking for RemoteDebug library in registry
Found: https://platformio.org/lib/show/1266/RemoteDebug
LibraryManager: Installing id=1266
Downloading...
Unpacking...
RemoteDebug @ 3.0.5 has been successfully installed!
Found 40 compatible libraries
Scanning dependencies...
Dependency Graph
|-- <Adafruit GFX Library> 1.5.7
|   |-- <SPI> 1.0
|-- <Adafruit SSD1351 library> 1.2.2
|   |-- <Adafruit GFX Library> 1.5.7
|   |   |-- <SPI> 1.0
|-- <HARestAPI> 0.2
|   |-- <ESP8266WiFi> 1.0
|-- <ArduinoJson> 6.12.0
|-- <RemoteDebug> 3.0.5
|   |-- <WiFi>
|   |   |-- <Colors>
|   |   |   |-- <Random>
|   |   |-- <ESP8266WiFi> 1.0
|   |-- <Ethernet(esp8266)> 1.0.4
|   |   |-- <SPI> 1.0
|   |-- <Hash> 1.0
|   |-- <SPI> 1.0
|   |-- <ESP8266WiFi> 1.0
|-- <OTAUpdates>
|   |-- <Adafruit SSD1351 library> 1.2.2
|   |   |-- <Adafruit GFX Library> 1.5.7
|   |   |   |-- <SPI> 1.0
|   |-- <Text>
|   |   |-- <Adafruit SSD1351 library> 1.2.2
|   |   |   |-- <Adafruit GFX Library> 1.5.7
|   |   |   |   |-- <SPI> 1.0
|   |   |-- <DLog>
|   |   |   |-- <RemoteDebug> 3.0.5
|   |   |   |   |-- <WiFi>
|   |   |   |   |   |-- <Colors>
|   |   |   |   |   |   |-- <Random>
|   |   |   |   |   |-- <ESP8266WiFi> 1.0
|   |   |   |   |-- <Ethernet(esp8266)> 1.0.4
|   |   |   |   |   |-- <SPI> 1.0
|   |   |   |   |-- <Hash> 1.0
|   |   |   |   |-- <SPI> 1.0
|   |   |   |   |-- <ESP8266WiFi> 1.0
|   |   |-- <Colors>
|   |   |   |-- <Random>
|   |-- <WiFi>
|   |   |-- <Colors>
|   |   |   |-- <Random>
|   |   |-- <ESP8266WiFi> 1.0
|   |-- <DLog>
|   |   |-- <RemoteDebug> 3.0.5
|   |   |   |-- <WiFi>
|   |   |   |   |-- <Colors>
|   |   |   |   |   |-- <Random>
|   |   |   |   |-- <ESP8266WiFi> 1.0
|   |   |   |-- <Ethernet(esp8266)> 1.0.4
|   |   |   |   |-- <SPI> 1.0
|   |   |   |-- <Hash> 1.0
|   |   |   |-- <SPI> 1.0
|   |   |   |-- <ESP8266WiFi> 1.0
|   |-- <Colors>
|   |   |-- <Random>
|   |-- <ArduinoOTA> 1.0
|   |   |-- <ESP8266mDNS> 1.2
|   |   |   |-- <ESP8266WiFi> 1.0
|   |   |-- <ESP8266WiFi> 1.0
|-- <WiFi>
|   |-- <Colors>
|   |   |-- <Random>
|   |-- <ESP8266WiFi> 1.0
|-- <DLog>
|   |-- <RemoteDebug> 3.0.5
|   |   |-- <WiFi>
|   |   |   |-- <Colors>
|   |   |   |   |-- <Random>
|   |   |   |-- <ESP8266WiFi> 1.0
|   |   |-- <Ethernet(esp8266)> 1.0.4
|   |   |   |-- <SPI> 1.0
|   |   |-- <Hash> 1.0
|   |   |-- <SPI> 1.0
|   |   |-- <ESP8266WiFi> 1.0
|-- <Colors>
|   |-- <Random>
|-- <Text>
|   |-- <Adafruit SSD1351 library> 1.2.2
|   |   |-- <Adafruit GFX Library> 1.5.7
|   |   |   |-- <SPI> 1.0
|   |-- <DLog>
|   |   |-- <RemoteDebug> 3.0.5
|   |   |   |-- <WiFi>
|   |   |   |   |-- <Colors>
|   |   |   |   |   |-- <Random>
|   |   |   |   |-- <ESP8266WiFi> 1.0
|   |   |   |-- <Ethernet(esp8266)> 1.0.4
|   |   |   |   |-- <SPI> 1.0
|   |   |   |-- <Hash> 1.0
|   |   |   |-- <SPI> 1.0
|   |   |   |-- <ESP8266WiFi> 1.0
|   |-- <Colors>
|   |   |-- <Random>
|-- <SPI> 1.0
Warning! We have just detected `upload_port` as IP address or host name of ESP device. `upload_protocol` is switched to `espota`.
Please specify `upload_protocol = espota` in `platformio.ini` project configuration file.
Compiling .pio/build/esp12e/src/main.cpp.o
Generating LD script .pio/build/esp12e/ld/local.eagle.app.v6.common.ld
Compiling .pio/build/esp12e/lib7b3/SPI/SPI.cpp.o
Archiving .pio/build/esp12e/lib7b3/libSPI.a
Indexing .pio/build/esp12e/lib7b3/libSPI.a
Compiling .pio/build/esp12e/libe1c/Adafruit GFX Library_ID13/Adafruit_GFX.cpp.o
Compiling .pio/build/esp12e/libe1c/Adafruit GFX Library_ID13/Adafruit_SPITFT.cpp.o
Compiling .pio/build/esp12e/libe1c/Adafruit GFX Library_ID13/glcdfont.c.o
Compiling .pio/build/esp12e/lib7f4/Adafruit SSD1351 library_ID266/Adafruit_SSD1351.cpp.o
Archiving .pio/build/esp12e/libe1c/libAdafruit GFX Library_ID13.a
Indexing .pio/build/esp12e/libe1c/libAdafruit GFX Library_ID13.a
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/BearSSLHelpers.cpp.o
Archiving .pio/build/esp12e/lib7f4/libAdafruit SSD1351 library_ID266.a
Indexing .pio/build/esp12e/lib7f4/libAdafruit SSD1351 library_ID266.a
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/CertStoreBearSSL.cpp.o
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/ESP8266WiFi.cpp.o
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/ESP8266WiFiAP.cpp.o
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/ESP8266WiFiGeneric.cpp.o
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/ESP8266WiFiMulti.cpp.o
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/ESP8266WiFiSTA-WPS.cpp.o
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/ESP8266WiFiSTA.cpp.o
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/ESP8266WiFiScan.cpp.o
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/WiFiClient.cpp.o
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/WiFiClientSecureAxTLS.cpp.o
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/WiFiClientSecureBearSSL.cpp.o
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/WiFiServer.cpp.o
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/WiFiServerSecureAxTLS.cpp.o
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/WiFiServerSecureBearSSL.cpp.o
Compiling .pio/build/esp12e/lib65f/ESP8266WiFi/WiFiUdp.cpp.o
Compiling .pio/build/esp12e/libe83/HARestAPI_ID5828/HARestAPI.cpp.o
Archiving .pio/build/esp12e/lib65f/libESP8266WiFi.a
Indexing .pio/build/esp12e/lib65f/libESP8266WiFi.a
Archiving .pio/build/esp12e/libe03/libArduinoJson_ID64.a
Indexing .pio/build/esp12e/libe03/libArduinoJson_ID64.a
Archiving .pio/build/esp12e/lib3ff/libRandom.a
Indexing .pio/build/esp12e/lib3ff/libRandom.a
Archiving .pio/build/esp12e/lib214/libColors.a
Indexing .pio/build/esp12e/lib214/libColors.a
Compiling .pio/build/esp12e/lib013/WiFi/WiFi.cpp.o
In file included from /home/travis/build/johnboiles/blocks-with-screens/blockware/lib/WiFi/WiFi.cpp:4:0:
/home/travis/build/johnboiles/blocks-with-screens/blockware/lib/Text/Text.h:2:30: fatal error: Adafruit_SSD1351.h: No such file or directory
**************************************************************************
* Looking for Adafruit_SSD1351.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:Adafruit_SSD1351.h"
* Web  > https://platformio.org/lib/search?query=header:Adafruit_SSD1351.h
*
**************************************************************************
 #include <Adafruit_SSD1351.h>
                              ^
compilation terminated.
*** [.pio/build/esp12e/lib013/WiFi/WiFi.cpp.o] Error 1
.pio/libdeps/esp12e/HARestAPI_ID5828/src/HARestAPI.cpp: In member function 'String HARestAPI::sendGetHA(String, String)':
.pio/libdeps/esp12e/HARestAPI_ID5828/src/HARestAPI.cpp:98:39: warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
     if(_debug & _fingerprint.length() > 0 ) {
                                       ^
.pio/libdeps/esp12e/HARestAPI_ID5828/src/HARestAPI.cpp:99:67: warning: 'bool BearSSL::WiFiClientSecure::verify(const char*, const char*)' is deprecated (declared at /home/travis/.platformio/packages/framework-arduinoespressif8266/libraries/ESP8266WiFi/src/WiFiClientSecureBearSSL.h:180): This is deprecated AxTLS API, check https://github.com/esp8266/Arduino/blob/master/libraries/ESP8266WiFi/src/WiFiClientSecure.h#L25-L99 [-Wdeprecated-declarations]
       if (wsclient->verify(_fingerprint.c_str(), _serverip.c_str())) {
                                                                   ^
.pio/libdeps/esp12e/HARestAPI_ID5828/src/HARestAPI.cpp: In member function 'bool HARestAPI::sendPostHA(String, String)':
.pio/libdeps/esp12e/HARestAPI_ID5828/src/HARestAPI.cpp:174:39: warning: suggest parentheses around comparison in operand of '&' [-Wparentheses]
     if(_debug & _fingerprint.length() > 0 ) {
                                       ^
.pio/libdeps/esp12e/HARestAPI_ID5828/src/HARestAPI.cpp:175:67: warning: 'bool BearSSL::WiFiClientSecure::verify(const char*, const char*)' is deprecated (declared at /home/travis/.platformio/packages/framework-arduinoespressif8266/libraries/ESP8266WiFi/src/WiFiClientSecureBearSSL.h:180): This is deprecated AxTLS API, check https://github.com/esp8266/Arduino/blob/master/libraries/ESP8266WiFi/src/WiFiClientSecure.h#L25-L99 [-Wdeprecated-declarations]
       if (wsclient->verify(_fingerprint.c_str(), _serverip.c_str())) {
                                                                   ^
========================= [FAILED] Took 17.34 seconds =========================
../lib/Makefile/common.mk:7: recipe for target 'test' failed
make[1]: *** [test] Error 1
make[1]: Leaving directory '/home/travis/build/johnboiles/blocks-with-screens/blockware/homeassistant-display'
Makefile:2: recipe for target 'test' failed
make: *** [test] Error 2
The command "make test" exited with 2.
```

**Comment by johnboiles** [comment]
Looks like that library wasn't included in the cpp file so the dependency manager wasn't able to figure out it was required.

```
|   |-- <WiFi>
|   |   |-- <Colors>
|   |   |   |-- <Random>
|   |   |-- <ESP8266WiFi> 1.0
```

I added back the header to the `cpp` file and we'll see if it fixes it

**Comment by johnboiles** [comment]
[Here's](https://stackoverflow.com/questions/1067226/c-multi-line-macro-do-while0-vs-scope-block) the reason for using do/while in macros.

It's basically to support `if` without a bracket:

`if (something) outputf(WHITE, 1, "test %s", someString);`

**Comment by magus** [comment]
Oh nice thanks for that link good to know

**Pull request merged**